jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py

```diff
@@ -1,128 +1,280 @@
-        traversal_paths = parameters.get(
-            'traversal_paths', self.default_traversal_paths
-        )
 
-        for doc in docs.traverse_flat(traversal_paths):
-            indices, dists = self._indexer.knn_query(doc.embedding, k=top_k)
-            for idx, dist in zip(indices[0], dists[0]):
-                match = Document(id=self._ids[idx], embedding=self._vecs[idx])
-                if self.is_distance:
-                    match.scores[self.metric] = dist
-                else:
-                    if self.metric == 'cosine' or self.metric == 'ip':
-                        match.scores[self.metric] = 1 - dist
-                    else:
-                        match.scores[self.metric] = 1 / (1 + dist)
-
-                doc.matches.append(match)
-
-    @requests(on='/fill_embedding')
-    def fill_embedding(self, docs: Optional[DocumentArray], **kwargs):
+        embeddings = docs_to_update.embeddings
+        if embeddings.shape[-1] != self.dim:
+            raise ValueError(
+                f'Attempted to index vectors with dimension'
+                f' {embeddings.shape[-1]}, but dimension of index is {self.dim}'
+            )
+
+        ids = docs_to_update.get_attributes('id')
+        index_size = self._index.element_count
+        doc_inds = []
+        for doc in docs_to_update:
+            if doc.id not in self._ids_to_inds:
+                doc_inds.append(index_size)
+                index_size += 1
+            else:
+                self.logger.info(
+                    f'Document with id {doc.id} already in index, updating.'
+                )
+                doc_inds.append(self._ids_to_inds[doc.id])
+
+        self._index.add_items(embeddings, ids=doc_inds)
+        self._ids_to_inds.update({_id: ind for _id, ind in zip(ids, doc_inds)})
+
+    @requests(on='/update')
+    def update(
+        self, docs: Optional[DocumentArray] = None, parameters: Dict = {}, **kwargs
+    ):
+        """Update the Documents' embeddings. If a Document is not already present in
+        the index, it will get ignored, and a warning will be raised.
+
+        :param docs: Documents whose `embedding` to update.
+        :param parameters: Dictionary with optional parameters that can be used to
+            override the parameters set at initialization. The only supported key is
+            `traversal_paths`.
+        """
+        traversal_paths = parameters.get('traversal_paths', self.traversal_paths)
         if docs is None:
             return
-        for doc in docs:
-            doc_idx = self._doc_id_to_offset.get(doc.id)
-            if doc_idx is not None:
-                doc.embedding = np.array(self._indexer.get_items([int(doc_idx)])[0])
+
+        docs_to_update = docs.traverse_flat(traversal_paths)
+        if len(docs_to_update) == 0:
+            return
+
+        doc_inds, docs_filtered = [], []
+        for doc in docs_to_update:
+            if doc.id not in self._ids_to_inds:
```

Comment:
I would say to store these in one list and only emit one warning for all of them together. If you do one by one, it can slow down the process

Reply:
Printing all ids together would not look good in the output. Then I would just mention the number of documents that were updated